pytorch_lightning/core/optimizer.py

```diff
@@ -0,0 +1,226 @@
+from argparse import Namespace
+from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence, Tuple, Union
+
+import torch
+from torch import ScriptModule, Tensor
+from torch.nn import Module
+from torch.optim.optimizer import Optimizer
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.callbacks import Callback
+from pytorch_lightning.core.grads import GradInformation
+from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks, ModelHooks
+from pytorch_lightning.core.memory import ModelSummary
+from pytorch_lightning.core.saving import ALLOWED_CONFIG_TYPES, PRIMITIVE_TYPES, ModelIO
+from pytorch_lightning.core.step_result import Result
+from pytorch_lightning.utilities import AMPType, rank_zero_warn
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.parsing import AttributeDict, collect_init_args, get_init_args
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils
+
+TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists()
+
+if TPU_AVAILABLE:
+    import torch_xla.core.xla_model as xm
+
+
+def do_nothing_closure():
+    return
+
+
+class LightningOptimizer(Optimizer):
+
+    """
+    This class is used to wrap the user optimizers and handle properly
+    the backward and optimizer_step logic across accelerators, AMP, accumulated_grad_batches
+    """
+
+    def __init__(self,
+                 optimizer: Optimizer = None,
+                 accumulate_grad_batches: Optional[int] = None):
+
+        if isinstance(accumulate_grad_batches, int) and accumulate_grad_batches < 1:
+            raise MisconfigurationException(f"accumulate_grad_batches parameters "
+                                            f"{accumulate_grad_batches} should be >= 1")
+
+        self._trainer = None
+        self._optimizer = optimizer
+        self._optimizer_idx = None
+        self._accumulate_grad_batches = accumulate_grad_batches
+        self._expose_optimizer_attr()
+
+    def _on_trainer_init(self, trainer, optimizer_idx, accumulate_grad_batches):
+        self._trainer = trainer
+        self._optimizer_idx = optimizer_idx
+        if self._accumulate_grad_batches is None:
+            self._accumulate_grad_batches = accumulate_grad_batches
+
+    @property
+    def accumulate_grad_batches(self):
+        return self._accumulate_grad_batches
+
+    @accumulate_grad_batches.setter
+    def accumulate_grad_batches(self, accumulate_grad_batches):
+        self._accumulate_grad_batches = accumulate_grad_batches
+
+    @property
+    def optimizer(self):
+        return self._optimizer
+
+    @optimizer.setter
+    def optimizer(self, optimizer):
+        return self._optimizer
+
+    def _expose_optimizer_attr(self):
+        for attr_name in dir(self._optimizer):
+            if ('__' in attr_name) or "step" == attr_name:
+                continue
```

Comment:
Should this really be silently discarded? Why not let the user hack it if they want to? They should know what they are doing.

Reply:
Removed the function.